xarray/core/ops.py

```diff
@@ -140,6 +140,30 @@ def array_equiv(arr1, arr2):
     return bool(flag_array.all())
 
 
+def array_notnull_equiv(arr1, arr2):
+    """Like np.array_equal, but also allows values to be NaN in either or both
+    arrays
+    """
+    arr1, arr2 = as_like_arrays(arr1, arr2)
+    if arr1.shape != arr2.shape:
+        return False
+
+    flag_array = (arr1 == arr2)
+
+    # GH837, GH861
```

Comment:
Yes it would be good not to repeat. However, if put in `isnull`, then `TypeError` might be suppressed when useful. Excepting the error I think can only work in situations (such as these two functions) where an array is being *modified* by the result and the whole operation can be ignored if fails. But maybe there is another way to do it?

The only quite ugly way I can think of is something like (leaving `isnull` as-is):
```python
def isnull_with_no_type_fail(a):
    # comments about why needed etc.
    try:
        b = isnull(a)
    except TypeError:
        b = None
    return b
```
and then in `array_equiv`/`array_notnull_equiv`

```python
...
arr1_isnull = isnull_with_no_type_fail(arr1)
if arr1_isnull is not None:
    arr2_isnull = isnull_with_no_type_fail(arr2)
    if arr2_isnull is not None:
        flag_array |= (arr1_isnull | arr2_isnull)
...
```